---
title: "Genetic diversity, inbreeding, and differentiation"
format:
  html:
    code-fold: true
---

The first figure focuses on changes in genetic diversity and differentiation
over time, and introduces the sampling area and focal species.

```{r}
#| label: load-packages
#| code-summary: Load up required packages and setup species and genome lists

library(data.table) # For reading in data while skipping columns
library(stringr) # For separating pop names into regions and sampling years
library(ggplot2) # For plotting
library(ggpubr) # For adding stats to graphs
library(cowplot) # For arranging plots into a multi-panel figure
library(ggtext) # For more control over axis text layout
library(dplyr) # For summarize function and piping
library(geosphere) # For including distances on F~ST~ plot
library(scales) # For some control of plot scales
library(png) # For reading in maps made in QGIS
library(grid) # For plotting maps made in QGIS with ggplot
library(knitr) # For markdown tables in report
library(dplyr) # For summary tables in groups

species <- c("picarus", "pargus", "csemiargus")
genomes <- c("ilPolIcar1.1", "ilPleArgu1.3", "ilCyaSemi1.1")
```

## Heterozygosity

For heterozygosity, read in the individual heterozygosity results from ANGSD.
These are genome-wide average estimates of heterozygous sites per 1000bp. They
also have a confidence interval for this from bootstraps of the single sample
SFS. Since this is mixing modern and historical samples with DNA damage, these
estimates are made excluding transitions. We test for a change in heterozygosity
across the time periods using a t-test (unequal variance) comparing all modern
samples to all historical samples within each species.

```{r}
#| label: tbl-heterozygosity

hz_all <- c()

hz_stats <- c()

for (i in c(1:length(species))) {
  s <- species[i]
  g <- genomes[i]
  for (d in c("", ".dp3")) {
    depth <- str_sub(d, -1)
    hz <- fread(
      paste0(
        "../angsd/results/datasets/historical-",
        s,
        "-notrans/analyses/heterozygosity/historical-",
        s,
        "-notrans.",
        g,
        "_all",
        d,
        "_allsites-filts_heterozygosity.tsv"
      ),
      header = TRUE,
      sep = "\t"
    )
    samples <- fread(
      paste0(
        "../angsd/results/datasets/historical-",
        s,
        "-notrans/qc/historical-",
        s,
        "-notrans.",
        g,
        "_all",
        d,
        ".sampleqc.tsv"
      ),
      header = TRUE,
      sep = "\t",
      select = c("sample", "population", "time", "allsites-filts.depth.mean")
    )
    hz$dp <- depth
    hz <- merge(hz, samples, by = "sample")
    if (depth != "") {
      hz <- hz[!hz$`allsites-filts.depth.mean` < as.integer(depth) - 0.1, ]
    }
    hz$species <- s
    testhistmod <- t.test(
      hz[hz$time == "historical", "heterozygous_per_1000bp"],
      hz[hz$time == "modern", "heterozygous_per_1000bp"]
    )
    stats <- c(
      s,
      depth,
      testhistmod$estimate[2] - testhistmod$estimate[1],
      -testhistmod$conf.int[2:1],
      testhistmod$method,
      testhistmod$statistic,
      testhistmod$p.value
    )
    hz_stats <- data.frame(rbind(hz_stats, stats))
    hz_all <- rbind(hz_all, hz)
  }
}


colnames(hz_stats) <- c(
  "Species", "Depth", "Mean Diff", "CI (Lower)", "CI (Upper)", "Method",
  "t-value", "p-value"
)

write.table(hz_stats, "../tables/hz_stats.tsv",
  row.names = FALSE, quote = FALSE, sep = "\t"
)

hz_all$year <- str_sub(hz_all$population, -4, -1)
hz_all$region <- str_sub(hz_all$population, 1, -5)

hz_all$species <- factor(
  hz_all$species,
  levels = species,
  labels = c("Po. icarus", "Pl. argus", "Cy. semiargus")
)

hz_all$time <- factor(
  hz_all$time,
  levels = c("historical", "modern"),
  labels = c("Historical", "Modern")
)

hz_all$region <- factor(
  hz_all$region,
  levels = c("WSkane", "SWSkane", "ESkane", "SESkane", "NSmaland", "Oland"),
  labels = c(
    "W Skåne", "SW Skåne", "E Skåne", "SE Skåne", "N Småland", "Öland"
  )
)

hz_table <- hz_all[, c(
  "sample", "species", "time", "pop", "dp",
  "heterozygous_per_1000bp", "lower_95_CI", "upper_95_CI"
)]

write.table(hz_table, "../tables/hz_ests.tsv",
  row.names = FALSE, quote = FALSE, sep = "\t"
)

kable(hz_table, row.names = FALSE)
```

Plot the heterozygosity by species. Show points colored by the sampled
areas, but group time periods together in a single boxplot.

```{r}
#| label: fig-het-box
#| fig-width: 6
#| fig-height: 2.5

hzbox <- ggplot(hz_all[hz_all$dp == "3", ], aes(x = time, y = heterozygous_per_1000bp)) +
  facet_grid(cols = vars(species)) +
  geom_boxplot(
    aes(x = time, y = heterozygous_per_1000bp),
    inherit.aes = FALSE,
    outliers = FALSE
  ) +
  geom_point(
    aes(color = region, shape = time),
    position = position_jitterdodge(jitter.width = 0.05)
  ) +
  scale_shape_manual(values = c(17, 16), guide = "none") +
  ylab("<span style='font-size: 9pt'>Heterozygous sites per 1000bp</span><br><span style='font-size: 7pt'>(excl. transitions)</span>") +
  scale_color_manual(
    values = c("#244F6E", "#244F6E", "#658399", "#658399", "#8A2A3B", "#fecd03"),
    name = "Geographic Region"
  ) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.20))) +
  geom_pwc(hide.ns = TRUE, label = "p.signif", method = "t.test") +
  theme_bw() +
  theme(
    legend.position.inside = c(0.08, 0.35),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y = element_markdown(),
    strip.text.x = element_text(size = 10, face = "italic")
  )

hzbox
```

## Nucleotide diversity, Watterson's θ, and Tajima's D

To examine the extent of within population genetic diversity changes, we will
look at nucleotide diversity (measuring the average pairwise differences
between sequences), Watterson's θ (measuring the number of segregating sites),
and Tajima's D (measuring excess or deficit of low frequency variation compared
to the neutral expectation). As nucleotide diversity is less impacted by rare
variants, sample size is less of a concern. Watterson's θ and Tajima's D,
however, are heavily influenced by rare variants and as such, can change
considerably as more samples are added. So, for these metrics, we will be
comparing the median of 100 subsampled estimates of 4 individuals each. In
general, we will consider populations to have different estimates if there is
no overlap in the confidence intervals of any of the individual subsamples.

First, we will read in the subsampled results. Each subsample has a mean for
each of the three metrics, as well as a confidence interval for each, built by
bootstrapping across genomic windows.

```{r}
#| label: suppfig-thetas
#| fig-width: 7
#| fig-height: 7

theta_ests <- c()

for (i in c(1:length(species))) {
  s <- species[i]
  g <- genomes[i]
  for (d in c("", ".dp3")) {
    # Have to read in a bit weird as I accidentally put all values on separate
    # lines of the data file in my script.
    theta <- readLines(
      paste0(
        "../angsd/results/datasets/historical-",
        s,
        "-notrans/analyses/thetas/downsampled/historical-",
        s,
        "-notrans.",
        g,
        "_all",
        d,
        ".downsampled_allsites-filts.thetaMean.10000_5000.tsv"
      )
    )
    header <- read.table(text = theta[1], sep = "\t", header = TRUE)
    theta <- as.data.frame(matrix(theta[-1], ncol = 12, byrow = TRUE))
    colnames(theta) <- colnames(header)
    theta$species <- s
    theta$dp <- d
    theta_ests <- rbind(theta_ests, theta)
  }
}

theta_ests[, -c(1, 13:14)] <- sapply(theta_ests[, -c(1, 13:14)], as.numeric)

theta_ests <- theta_ests %>%
  group_by(species, dp, pop) %>%
  summarise(
    med_pi = median(pi),
    min_pi = min(pi),
    max_pi = max(pi),
    med_watt = median(watterson),
    min_watt = min(watterson),
    max_watt = max(watterson),
    med_taj = median(tajima),
    min_taj = min(tajima),
    max_taj = max(tajima)
  )

write.table(theta_ests, "../tables/theta_ests.tsv",
  sep = "\t", quote = FALSE,
  row.names = FALSE
)

theta_stats <- c()

for (i in c(1:length(species))) {
  s <- species[i]
  g <- genomes[i]
  for (d in c("", ".dp3")) {
    theta <- read.table(
      paste0(
        "../angsd/results/datasets/historical-",
        s,
        "-notrans/analyses/thetas/downsampled/historical-",
        s,
        "-notrans.",
        g,
        "_modern-historical",
        d,
        ".N4_allsites-filts.wilcoxon.tsv"
      ),
      header = TRUE,
      sep = "\t"
    )
    # fix a coding mistake that added the depth subsampling wildcard to the
    # historical population
    theta$histpop <- sapply(
      theta$histpop,
      function(x) str_split(x, "\\.")[[1]][1]
    )
    theta$species <- s
    theta$dp <- d
    theta_stats <- rbind(theta_stats, theta)
  }
}

theta_stats$modyear <- str_sub(theta_stats$modpop, -4, -1)
theta_stats$modreg <- str_sub(theta_stats$modpop, 1, -5)
theta_stats$histyear <- str_sub(theta_stats$histpop, -4, -1)
theta_stats$histreg <- str_sub(theta_stats$histpop, 1, -5)

theta_stats$species <- factor(
  theta_stats$species,
  levels = species,
  labels = c("Po. icarus", "Pl. argus", "Cy. semiargus")
)

theta_stats$modreg <- factor(
  theta_stats$modreg,
  levels = c("WSkane", "SWSkane", "ESkane", "SESkane", "NSmaland", "Oland"),
  labels = c(
    "W Skåne", "SW Skåne", "E Skåne", "SE Skåne", "Småland", "Öland"
  )
)

theta_stats$histreg <- factor(
  theta_stats$histreg,
  levels = c("WSkane", "SWSkane", "ESkane", "SESkane", "NSmaland", "Oland"),
  labels = c(
    "W Skåne", "SW Skåne", "E Skåne", "SE Skåne", "Småland", "Öland"
  )
)

theta_stats$dp <- factor(
  theta_stats$dp,
  levels = c("", ".dp3"),
  labels = c("Full Depth", "3X Depth")
)

theta_stats$xlab <- paste0(
  theta_stats$modreg, " (", theta_stats$modyear, ") -<br>",
  theta_stats$histreg, " (", theta_stats$histyear, ")"
)

colorscale <- c("#244F6E", "#658399", "#658399", "#8A2A3B", "#fecd03")

pi <- ggplot(theta_stats, aes(x = xlab, y = pi.effsize)) +
  facet_grid(cols = vars(species), scales = "free") +
  geom_violin(
    aes(fill = modreg, linetype = dp)
  ) +
  geom_hline(yintercept = 0) +
  ylim(c(-1, 1)) +
  ylab("Δ Nucleotide Diversity<br>Effect Size") +
  scale_fill_manual(
    values = colorscale,
    name = "Geographic Region"
  ) +
  theme_bw() +
  theme(
    # legend.position.inside = c(0.08, 0.35),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y = element_markdown(),
    strip.text.x = element_text(size = 10, face = "italic")
  )

watterson <- ggplot(theta_stats, aes(x = xlab, y = watterson.effsize)) +
  facet_grid(cols = vars(species), scales = "free") +
  geom_violin(
    aes(fill = modreg, linetype = dp)
  ) +
  geom_hline(yintercept = 0) +
  ylim(c(-1, 1)) +
  ylab("Δ Watterson's *θ*<br>Effect Size") +
  scale_fill_manual(
    values = colorscale,
    name = "Geographic Region"
  ) +
  theme_bw() +
  theme(
    # legend.position.inside = c(0.08, 0.35),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y = element_markdown(),
    strip.text.x = element_text(size = 10, face = "italic")
  )

tajima <- ggplot(theta_stats, aes(x = xlab, y = tajima.effsize)) +
  facet_grid(cols = vars(species), scales = "free") +
  geom_violin(
    aes(fill = modreg, linetype = dp)
  ) +
  geom_hline(yintercept = 0) +
  ylim(c(-1, 1)) +
  ylab("Δ Tajima's *D*<br>Effect Size") +
  scale_fill_manual(
    values = colorscale,
    guide = "none"
  ) +
  theme_bw() +
  theme(
    # legend.position.inside = c(0.08, 0.35),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y = element_markdown(),
    strip.text.x = element_text(size = 10, face = "italic")
  )

plot_grid(
  pi + theme(
    axis.text.x = element_blank(),
    legend.position = "none"
  ),
  watterson + theme(
    axis.text.x = element_blank(),
    legend.position = "none"
  ),
  tajima + theme(
    axis.text.x = element_markdown(size = 6, angle = 45, vjust = 1, hjust = 1),
    legend.position = "bottom",
    legend.title = element_blank()
  ),
  nrow = 3, ncol = 1,
  rel_heights = c(1, 1, 1.7),
  labels = c("A", "B", "C")
)

ggsave("../figures/suppfig-thetas.png", dpi = 600, create.dir = TRUE)

theta_stats$pair <- paste0(theta_stats$modpop, "_", theta_stats$histpop)

# theta_stats["pi.pval"][theta_stats["pi.pval"] == 0] <- 2.22e-16

theta_table <- theta_stats %>%
  group_by(species, pair, modreg, modyear, histreg, histyear, dp) %>%
  summarise(
    med_pidiff = median(pi.diff),
    med_piret = median(pi.retained),
    med_piz = median(pi.z),
    med_pieff = median(pi.effsize),
    med_watdiff = median(watterson.diff),
    med_watret = median(watterson.retained),
    med_watz = median(watterson.z),
    med_wateff = median(watterson.effsize),
    med_tajdiff = median(tajima.diff),
    med_tajret = median(tajima.retained),
    med_tajz = median(tajima.z),
    med_tajeff = median(tajima.effsize)
  )

theta_table$elapsed <- as.numeric(theta_table$modyear) - as.numeric(theta_table$histyear)
theta_table$pi.ret.peryr <- exp(
  log(theta_table$med_piret) / theta_table$elapsed
)
theta_table$wat.ret.peryr <- exp(
  log(theta_table$med_watret) / theta_table$elapsed
)

theta_table$pi.ret.100yr <- (theta_table$pi.ret.peryr^100) * 100
theta_table$wat.ret.100yr <- (theta_table$wat.ret.peryr^100) * 100

write.table(theta_table, "../tables/theta_stats.tsv",
  row.names = FALSE, quote = FALSE, sep = "\t"
)
```

## Inbreeding coefficients

Next, we read in the inbreeding coefficients calculated from runs of
homozygosity > 100kb in each sample. For these, genotypes are called per sample,
only including genotypes with sequencing depth >=6 and an allelic balance from
0.2-0.8 for heterozygous genotypes. Only samples with >5X average sequencing
depth in the filtered region are included.

Each species then has all samples merged, removing variants fixed across the
dataset, as well as SNPs within 5bp of indels along with indels themselves.
Finally, to account for DNA damage, all transition variants are removed. These
variants are then run through bcftools roh to estimate IBD tracts, using a
rough species specific recombination rate (assuming 50cM per chr), setting a
default minor allele frequency, and ignoring homozygous reference genotypes.
This pattern of filtering mimics how bcftools would handle a single sample roh
analysis, but with the benefit of allowing removal of variants fixed in the
dataset as homozygous alternate. These variants can skew roh estimation, as they
likely are not segregating in the sample area, but will be treated as having
a default minor allele frequency of 0.4, which would create a bias in the model
based on how diverged the sample set is from the reference.

```{r}
#| label: read-froh

froh_all <- c()

for (i in c(1:length(species))) {
  s <- species[i]
  g <- genomes[i]
  froh <- fread(
    paste0(
      "../angsd/results/datasets/historical-",
      s,
      "-notrans/plots/inbreeding/historical-",
      s,
      "-notrans.",
      g,
      "_all.dp5_allsites-filts.filtered_mindp4-biallelic.notrans.fmiss0.2.bcftools.ind_froh.tsv"
    ),
    header = TRUE,
    sep = "\t",
    select = c("sample", "population", "time", "depth", "Froh")
  )
  froh$species <- s
  froh_all <- rbind(froh_all, froh)
}

froh_all$year <- str_sub(froh_all$population, -4, -1)
froh_all$region <- str_sub(froh_all$population, 1, -5)

froh_all$species <- factor(
  froh_all$species,
  levels = species,
  labels = c("Po. icarus", "Pl. argus", "Cy. semiargus")
)

froh_all$time <- factor(
  froh_all$time,
  levels = c("historical", "modern"),
  labels = c("Historical", "Modern")
)

froh_all$region <- factor(
  froh_all$region,
  levels = c("WSkane", "SWSkane", "ESkane", "SESkane", "NSmaland", "Oland"),
  labels = c("W Skåne", "SW Skåne", "E Skåne", "SE Skåne", "N Småland", "Öland")
)

froh_stats <- c()

for (i in c(1:length(species))) {
  s <- c("Po. icarus", "Pl. argus", "Cy. semiargus")[i]
  mod <- froh_all[froh_all$species == s & froh_all$time == "Modern", ]$Froh
  hist <- froh_all[froh_all$species == s & froh_all$time == "Historical", ]$Froh
  stat <- t.test(mod, hist)
  row <- c(
    s,
    stat$estimate[1],
    stat$estimate[2],
    stat$method,
    stat$estimate[1] - stat$estimate[2],
    stat$conf.int,
    stat$statistic,
    stat$p.value
  )
  froh_stats <- rbind(froh_stats, row)
}

colnames(froh_stats) <- c(
  "Species", "Mean Froh >100kb (Modern)", "Mean Froh >100kb (Historical)",
  "Method", "Mean change Froh (Modern-Historical)", "CI (Lower)", "CI (Upper)",
  "t-value", "p-value"
)

write.table(froh_stats,
  file = "../tables/froh_stats.tsv", quote = FALSE,
  sep = "\t", row.names = FALSE
)
```

Plot these the same way as heterozygosity.

```{r}
#| label: fig-froh-box
#| fig-width: 6
#| fig-height: 2.5

frohbox <- ggplot(froh_all, aes(x = time, y = Froh)) +
  facet_grid(cols = vars(species)) +
  geom_boxplot(
    aes(x = time, y = Froh),
    inherit.aes = FALSE,
    outliers = FALSE
  ) +
  geom_point(
    aes(color = region, shape = time),
    position = position_jitterdodge(jitter.width = 0.05)
  ) +
  scale_shape_manual(values = c(17, 16), guide = "none") +
  ylab("<span style='font-size: 9pt'>*F~RoH~* </span><span style='font-size: 7pt'>(*RoH* > 100kb)</span>") +
  scale_color_manual(
    values = c("#244F6E", "#244F6E", "#658399", "#658399", "#8A2A3B", "#fecd03"),
    name = "Geographic Region"
  ) +
  scale_y_continuous(expand = expansion(mult = c(0.05, 0.20))) +
  geom_pwc(hide.ns = TRUE, label = "p.signif", method = "t.test") +
  theme_bw() +
  theme(
    legend.position.inside = c(0.08, 0.35),
    panel.grid = element_blank(),
    axis.title.x = element_blank(),
    axis.title.y = element_markdown(),
    strip.text.x = element_text(size = 10, face = "italic")
  )

frohbox
```


## Population differention using *F~ST~*

As a measure of population differentiation, we use pairwise population *F*~ST~.
This is estimated in ANGSD using the Bhatia estimator, which is more robust to
lower sample sizes. We compare these values within species, visualizing as a
difference in *F*~ST~ between equivalent historical and modern population pairs,
including their geographic distance in the visualization.

```{r}
#| label: tbl-fst

fst_all <- c()

# read in F_ST results for population pairs
for (i in c(1:length(species))) {
  s <- species[i]
  g <- genomes[i]
  for (d in c("", ".dp3")) {
    fst <- fread(
      paste0(
        "../angsd/results/datasets/historical-",
        s,
        "-notrans/analyses/fst/historical-",
        s,
        "-notrans.",
        g,
        "_poppairs",
        d,
        "_allsites-filts.fst.global.tsv"
      ),
      header = TRUE,
      sep = "\t",
      select = c("pop1", "pop2", "unweight.fst", "weight.fst")
    )
    fst$pop2 <- sapply(
      fst$pop2,
      function(x) str_split(x, "\\.")[[1]][1]
    )
    fst$species <- s
    fst$dp <- d
    fst_all <- rbind(fst_all, fst)
  }
}

fst <- fst_all

# Read in sampling info, including coordinates
regions <- read.table("../config/pop_groups.tsv", header = TRUE)

# Merge sampling info with F_ST data
fst <- merge(fst, regions, by.x = "pop1", by.y = "population")
fst <- merge(fst, regions, by.x = "pop2", by.y = "population")

# Get rid of pairs within same time period, get pair 'name', and set time
# categories for pair
fst <- fst[abs(fst$year.x - fst$year.y) < 60, ]
fst$pair <- paste0(fst$region.x, "-", fst$region.y)
fst$time <- "placeholder"
fst[fst$year.x < 2000, ]$time <- "historical"
fst[fst$year.x > 2000, ]$time <- "modern"
# Calculate distance between population pairs
fst$dist <- as.vector(
  apply(
    fst[, c("lon.x", "lat.x", "lon.y", "lat.y")], 1, function(x) distm(c(x[1], x[2]), c(x[3], x[4]), fun = distGeo)
  )
)
# Set F_ST < 0 to 0 (this comes from the optimization in ANGSD and doesn't have
# any biological meaning different than 0)
fst$weight.fst[fst$weight.fst < 0] <- 0
# Remove pairs where populations contain single samples
fst <- fst[!fst$pop2 %in% c("ESkane1917", "WSkane1943"), ]
fst <- fst[!fst$pop1 %in% c("ESkane1917", "WSkane1943"), ]

# Set up variables for plotting
fst$species <- factor(
  fst$species,
  levels = c("picarus", "pargus", "csemiargus"),
  labels = c("Po. icarus", "Pl. argus", "Cy. semiargus")
)
fst$time <- factor(
  fst$time,
  levels = c("historical", "modern"),
  labels = c("Historical", "Modern")
)

write.table(fst,
  file = "../tables/fst_ests.tsv", sep = "\t",
  quote = FALSE, row.names = FALSE
)

fst_stats <- c()

for (i in c(1:length(species))) {
  s <- c("Po. icarus", "Pl. argus", "Cy. semiargus")[i]
  for (d in c("", ".dp3")) {
    hist <- fst[fst$species == s & fst$time == "Historical" & fst$dp == d, ]
    mod <- fst[fst$species == s & fst$time == "Modern" & fst$dp == d, ]
    if ((nrow(hist) > 1) & nrow(mod) > 1) {
      unpaired <- t.test(mod$weight.fst, hist$weight.fst)
      unpairrow <- c(
        s,
        d,
        "All population pairs",
        unpaired$method,
        unpaired$estimate[1] - unpaired$estimate[2],
        unpaired$conf.int,
        unpaired$statistic,
        unpaired$p.value
      )
    } else {
      unpairrow <- c(
        s,
        d,
        "All population pairs",
        paste0(
          "Insufficient population pairs (Modern = ",
          nrow(mod),
          ", Historical = ",
          nrow(hist),
          ")"
        ),
        mean(mod$weight.fst) - mean(hist$weight.fst),
        NA,
        NA,
        NA,
        NA
      )
    }
    fst_stats <- rbind(fst_stats, unpairrow)
    df <- merge(mod, hist, by = "pair")
    if (nrow(df) > 1) {
      paired <- t.test(df$weight.fst.x, df$weight.fst.y, paired = TRUE)
      pairrow <- c(
        s,
        d,
        "Pairs in both mod and hist",
        paired$method,
        paired$estimate,
        paired$conf.int,
        paired$statistic,
        paired$p.value
      )
    } else {
      pairrow <- c(
        s,
        d,
        "Pairs in both mod and hist",
        paste0(
          "Insufficient population pairs in both time periods (N = ",
          nrow(df),
          ")"
        ),
        mean(df$weight.fst.x) - mean(df$weight.fst.y),
        NA,
        NA,
        NA,
        NA
      )
    }
    fst_stats <- rbind(fst_stats, pairrow)
  }
}

colnames(fst_stats) <- c(
  "Species", "Depth", "Pairing", "Method",
  "Mean Difference (Modern-Historical)", "CI (lower)", "CI (upper)", "t-value",
  "p-value"
)

write.table(fst_stats,
  file = "../tables/fst_stats.tsv", sep = "\t",
  row.names = FALSE, quote = FALSE
)
```

```{r}
#| label: fig-fst-box
#| fig-width: 6
#| fig-height: 2.5

fstplot <- ggplot(
  fst[fst$dp == ".dp3", ],
  aes(
    x = time,
    y = weight.fst,
    group = pair,
    col = dist / 1000,
    label = paste0(round(dist / 1000, 1), " km")
  )
) +
  geom_line(linewidth = 0.5) +
  geom_point(aes(shape = time), size = 2.5) +
  scale_shape_manual(values = c(17, 16), guide = "none") +
  # geom_text_repel() +
  facet_grid(cols = vars(species)) +
  theme_bw() +
  scale_color_gradient(
    low = "#55AFF4", high = "#132C44",
    name = "Distance (km)"
  ) +
  ylab("<span style='font-size: 9pt'>Weighted *F~ST~*</span>") +
  scale_y_continuous(
    labels = number_format(accuracy = 0.01),
    limits = c(0, 0.12)
  ) +
  guides(
    colour = guide_colourbar(position = "inside")
  ) +
  theme(
    legend.background = element_rect(color = "black", size = 0.2),
    legend.title = element_text(size = 8),
    legend.text = element_text(size = 7),
    legend.key.size = unit(0.4, "cm"),
    legend.justification.inside = c(0.02, 0.9),
    axis.title.x = element_blank(),
    panel.grid = element_blank(),
    axis.title.y = element_markdown(),
    strip.text = element_text(face = "italic", size = 10)
  )

fstplot
```


## Make Figure 1

We use cowplot in R to arrange these figures to produce a final figure for the
manuscript. This figure also includes a map prepared in QGIS and photos I've
taken of the butterflies.

```{r}
#| label: fig-diversity
#| fig-width: 10.5
#| fig-height: 7

arrhz <- hzbox +
  theme(
    legend.position = "none"
    # axis.text.x = element_blank(),
    # plot.margin = unit(c(0, 0, 0, 0), "cm")
  )

arrfroh <- frohbox +
  theme(
    legend.position = "none"
    # axis.text.x = element_blank(),
    # strip.text.x = element_blank(),
    # strip.background = element_blank(),
    # plot.margin = unit(c(0, 0, 0, 0), "cm")
  )

arrfst <- fstplot

map <- readPNG("../resources/gis/sampling_map/historical_map.png")
map <- rasterGrob(map, interpolate = TRUE)
map <- ggplot() +
  geom_blank() +
  annotation_custom(map, xmin = -Inf, xmax = Inf, ymin = -Inf, ymax = Inf) +
  theme(
    panel.background = element_rect(
      fill = "transparent", color = NA
    ),
    plot.background = element_rect(
      fill = "transparent", color = NA
    )
  )

butterflies <- readPNG("../images/butterfly_species_photos.png")
butterflies <- rasterGrob(butterflies, interpolate = TRUE)
butterflies <- ggplot() +
  geom_blank() +
  annotation_custom(
    butterflies,
    xmin = -Inf, xmax = Inf, ymin = -Inf, ymax = Inf
  ) +
  theme(
    panel.background = element_rect(
      fill = "transparent", color = NA
    ),
    plot.background = element_rect(
      fill = "transparent", color = NA
    )
  )

map_photos <- plot_grid(
  butterflies, map,
  labels = c("A", "B"),
  nrow = 2,
  ncol = 1,
  rel_heights = c(0.4, 1)
)

plot_grid(map_photos,
  plot_grid(arrhz, arrfroh, arrfst, ncol = 1, nrow = 3, labels = c("C", "D", "E"), align = "v", axis = "lr"),
  ncol = 2, nrow = 1
) + bgcolor("white")

ggsave("../figures/fig-diversity.png", dpi = 600, create.dir = TRUE)
```